ci/scan_flattened_deps.py

```diff
@@ -0,0 +1,313 @@
+    deps_content = file.read()
+
+    # Eval the content.
+    exec(deps_content, global_scope, local_scope)
+
+    # Extract the deps and filter.
+    deps_list = local_scope.get('vars')
+  queries = []  # list of queries to submit in bulk request to OSV API
+  with open(deps_flat_file, 'r') as file:
+    lines = file.readlines()
+
+  headers = {
+      'Content-Type': 'application/json',
+  }
+  osv_url = 'https://api.osv.dev/v1/querybatch'
+
+  if not os.path.exists(DEP_CLONE_DIR):
+    os.mkdir(DEP_CLONE_DIR)  #clone deps with upstream into temporary dir
+
+  # Extract commit hash, save in dictionary
+  for line in lines:
+    dep = line.strip().split(
+        '@'
+    )  # separate fully qualified dep into name + pinned hash
+
+    common_commit = get_common_ancestor_commit(dep, deps_list)
+    if isinstance(common_commit, str):
+      queries.append({'commit': common_commit})
+    else:
+      failed_deps.append(dep[0])
+
+  print(
+      'Dependencies that could not be parsed for ancestor commits: ' +
+      ', '.join(failed_deps)
+  )
+  try:
+    # clean up cloned upstream dependency directory
+    shutil.rmtree(
+        DEP_CLONE_DIR
+    )  # use shutil.rmtree since dir could be non-empty
+  except OSError as clone_dir_error:
+    print(
+        'Error cleaning up clone directory: %s : %s' %
+        (DEP_CLONE_DIR, clone_dir_error.strerror)
+    )
+  # Query OSV API using common ancestor commit for each dep
+  # return any vulnerabilities found.
+  data = json.dumps({'queries': queries}).encode('utf-8')
+  req = request.Request(osv_url, data, headers=headers)
+  with request.urlopen(req) as resp:
+    res_body = resp.read()
+    results_json = json.loads(res_body.decode('utf-8'))
+    if resp.status != 200:
+      print('Request error')
+    elif results_json['results'] == [{}]:
+      print('Found no vulnerabilities')
+    else:
+      results = results_json['results']
+      filtered_results = list(filter(lambda vuln: vuln != {}, results))
+      if len(filtered_results) > 0:
+        print(
+            'Found vulnerability on {vuln_count} dependenc(y/ies), adding to report'
+            .format(vuln_count=str(len(filtered_results)))
+        )
+        print(*filtered_results)
+        return filtered_results
+      print('Found no vulnerabilities')
+  return {}
+
+
+def get_common_ancestor_commit(dep, deps_list):
+  """
+  Given an input of a mirrored dep,
+  compare to the mapping of deps to their upstream
+  in DEPS and find a common ancestor
+  commit SHA value.
+
+  This is done by first cloning the mirrored dep,
+  then a branch which tracks the upstream.
+  From there,  git merge-base operates using the HEAD
+  commit SHA of the upstream branch and the pinned
+  SHA value of the mirrored branch
+  """
+  # dep[0] contains the mirror repo
+  # dep[1] contains the mirror's pinned SHA
+  # upstream is the origin repo
+  dep_name = dep[0].split('/')[-1].split('.')[0]
+  if UPSTREAM_PREFIX + dep_name in deps_list:
```

Comment:
optional nit: flip the sense, do an early return, then un-indent the `try: catch:`:
```python
if UPSTREAM_PREFIX + dep_name not in deps_list:
  print('did not find dep: ' + dep_name)
  return {}

try:
  ...
except ...:
  ...

return {}
```

Reply:
Thank you for this recommendation! I added this early return.